packages/@sanity/form-builder/src/inputs/ObjectInput/ConditionalField.tsx

```diff
@@ -0,0 +1,72 @@
+import React from 'react'
+import {BaseSchemaType, HiddenPredicateContext, HiddenOptionPredicate} from '@sanity/types'
+import {combineLatest, Observable, of} from 'rxjs'
+import {rxComponent} from 'react-rx'
+import {distinctUntilChanged, map, switchMap} from 'rxjs/operators'
+import withDocument from '../../utils/withDocument'
+
+export type HiddenOption = BaseSchemaType['hidden']
+
+function isThenable(value: any) {
+  return typeof value?.then === 'function'
+}
+
+function checkCondition(
+  schemaHiddenOption: HiddenOptionPredicate | undefined | boolean,
+  context: HiddenPredicateContext
+): boolean {
+  const result =
+    typeof schemaHiddenOption === 'function' ? schemaHiddenOption(context) : schemaHiddenOption
+
+  if (isThenable(result)) {
+    console.warn(
+      '[Warning]: The hidden option is either a promise or a promise returning function. Async callbacks for `hidden` option is not currently supported.'
+    )
+    return false
+  }
+  return Boolean(result)
+}
+
+interface Props {
+  hidden: HiddenOption
+  parent: Record<string, unknown> | undefined
+  value: unknown
+  children?: React.ReactNode
+}
+
+export const ConditionalField = (props: Props) => {
+  if (!props.hidden || props.hidden === true) {
+    return <>{props.hidden === true ? null : props.children}</>
+  }
+  return <ConditionalFieldWithDocument {...props} />
+}
+
+const ConditionalFieldWithDocument = withDocument(
+  rxComponent(function ConditionalFieldWithDocument(props$: Observable<Props & {document: any}>) {
```

Comment:
Wondering is this could be solved using `useMemo` somehow. Would probably be simpler to read and maintain for more people.

Reply:
Could `{document: any}` here instead be the interface imported from `withDocument`?

```tsx
import {WithDocumentProps} from '../../utils/withDocument'

// ...

const ConditionalFieldWithDocument = withDocument(
  rxComponent(function ConditionalFieldWithDocument(props$: Observable<Props & WithDocumentProps>) {
```

Reply:
> Wondering is this could be solved using useMemo somehow. Would probably be simpler to read and maintain for more people.

I agree. unless i'm mistaken, i think it would look like this

```js
const ConditionalFieldWithDocument = withDocument(({hidden, document, parent, value, children}) => {
  const shouldHide = useMemo(() => {
    return hidden({document, parent, value, path})
  }, [hidden, document, parent, value, path])

  return shouldHide ? null : children
})

ConditionalFieldWithDocument.displayName = 'ConditionalFieldWithDocument'
```

Reply:
Yeah, that’s much simpler IMO @ricokahler 👍 

Later, the `useMemo` can be refactored to a hook and support promises and observables, something like:
```tsx
function getHidden$(hidden, context) {
  return of(context).pipe(
    switchMap(c => {
      const val = hidden(c)

      if (!val || val === tue) return of(val)

      return from(val)
    })
  )
}

function useHiddenAsync(hidden, context) {
  const hidden$ = useMemo(() => getHidden$(hidden, context), [hidden, context])

  return useObservable(hidden$, false)
}

const ConditionalFieldWithDocument = withDocument(({hidden, document, parent, value, children}) => {
  const context = useMemo(
    () => ({document, parent, value, children}),
    [document, parent, value, children]
  )

  const shouldHide = useHiddenAsync(hidden, context)

  return shouldHide ? null : children
})

ConditionalFieldWithDocument.displayName = 'ConditionalFieldWithDocument'
```

Reply:
> Later, the useMemo can be refactored to a hook and support promises and observables, something like:

Agree in general, although this isn't quite what we want from the observable case though. If you provide a reactive/observable callback you should be able to build on a stream of context args by providing a callback that is called once initially with the context stream:

```ts
{
  //...
  hidden: {
    stream: (context$) =>
      context$.pipe(
        debounceTime(200),
        switchMap((context) => fetch(context.parent.productId)),
        map((product) => product.expired)
      )
  }
}
```

Reply:
I've refactored to functional component and it became a lot simpler, thanks for pointing out!